database/update_db.py

```diff
@@ -0,0 +1,73 @@
+import gspread
+from google.oauth2.service_account import Credentials
+from dotenv import load_dotenv;
+import json
+import os
+import redis
+import time
+
+config_path = 'config/default.json'
+
+# Load the configuration file
+with open(config_path, 'r') as config_file:
+    config = json.load(config_file)
+
+SCOPES = config['spreadsheet']['scopes']
+HOST = config['server']['host']
+PORT = config['server']['port']
+DB = config['server']['db']
+SHEETNAME = config['spreadsheet']['sheetname']
+WORKSHEET = config['spreadsheet']['worksheet']
+CATEGORYCOL = config['spreadsheet']['keyvals']['categorycol']
+CATEGORYROW = config['spreadsheet']['keyvals']['categoryrow']
+CONCEPTSCOL = config['spreadsheet']['keyvals']['conceptscol']
+CONCEPTSROW = config['spreadsheet']['keyvals']['conceptsrow']
+MAXPOINTSROW = config['spreadsheet']['keyvals']['maxpointsrow']
+MAXPOINTSCOL = config['spreadsheet']['keyvals']['maxpointscol']
+
+#needs both spreadsheet and drive access or else there is a permissions error, added as a viewer on the spreadsheet
+load_dotenv()
+credentials_json = os.getenv("GRADE_DB_SECRET")
+credentials_dict = json.loads(credentials_json)
+credentials = Credentials.from_service_account_info(credentials_dict, scopes=SCOPES)
+client = gspread.authorize(credentials)
+
+#redis setup
+redis_client = redis.Redis(host=HOST, port=PORT, db=DB) 
+
+def update_redis():
+    sheet = client.open(SHEETNAME).get_worksheet(WORKSHEET)
+    categories = sheet.row_values(CATEGORYROW)[CATEGORYCOL:] #gets the categories from row 2, starting from column C
+    concepts = sheet.row_values(CONCEPTSROW)[CONCEPTSCOL:] #gets the concepts from row 1, starting from column C
+    max_points = sheet.row_values(MAXPOINTSROW)[MAXPOINTSCOL:] #gets the max points from row 3, starting from column C
+    
+    category_scores = {}
+    for category, concept, points in zip(categories, concepts, max_points):
+        if category not in category_scores:
+            category_scores[category] = {} #creates a hashmap entry for each category
+        category_scores[category][concept] = points #nested hashmap of     category:concept:points
+    
+    redis_client.set("Categories", json.dumps(category_scores)) #the one record that holds all of the categories info
+    
+    records = sheet.get_all_records()
+    
+    for record in records:
```

Comment:
`O(|records| * max(|categories|, |concepts|) * |users_to_assignments|)` is reallllly inefficient.  Is there any way to speed this up?

Reply:
I don't think so...we have to iterate over each of the records, and for each of those records, based on how we want it formatted, we also need to iterate over each of the categories/concepts. We need to also make a dictionary entry for each category/concept for each user to hold their score.